File: frameworks/java/spring-boot/Implementation/src/main/java/implementation/routes/ContractRoutes.java
package implementation.routes;

import java.lang.management.ManagementFactory;

import org.apache.catalina.util.ServerInfo;
import org.springframework.boot.SpringBootVersion;
import org.springframework.boot.web.server.context.WebServerInitializedEvent;
import org.springframework.context.event.EventListener;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;
import tools.jackson.databind.json.JsonMapper;

/** /health and /__meta, which the contract asks of every framework outside the corpus. */
@RestController
public class ContractRoutes {

    /** /__meta: what ran, recorded on the result row and never checked. */
    public record Meta(String framework, String version, String runtime, String adapter, String serializer, Long bootMs) {}

    private final String serializer;

    /** Milliseconds from the start of the JVM to Tomcat listening. */
    private volatile Long bootMs;

    ContractRoutes(JsonMapper json) {
        this.serializer = "Jackson " + json.version();
    }

    @EventListener
    void listening(WebServerInitializedEvent event) {
        bootMs = System.currentTimeMillis() - ManagementFactory.getRuntimeMXBean().getStartTime();
    }

    // The payloads are loaded before Tomcat starts, so a server that answers has them.
    @GetMapping(value = "/health", produces = MediaType.TEXT_PLAIN_VALUE)
    public String health() {
        return "ok";
    }

    @GetMapping("/__meta")
    public Meta meta() {
        return new Meta("Spring Boot", SpringBootVersion.getVersion(), "Java " + Runtime.version(),
                ServerInfo.getServerInfo().replace('/', ' '), serializer, bootMs);
    }
}

File: frameworks/dotnet/carter/Implementation/Routes/ContractRoutes.cs
using System.Reflection;
using System.Runtime.InteropServices;
using Carter;

namespace Implementation.Routes;

/// <summary>/health and /__meta, which the contract asks of every target outside the corpus.</summary>
public sealed class ContractRoutes : ICarterModule
{
    public void AddRoutes(IEndpointRouteBuilder app)
    {
        // The payloads are loaded before the server starts, so a server that answers has them.
        app.MapGet("/health", () => Results.Text("ok"));

        app.MapGet("/__meta", () => new Meta("Carter", Version(typeof(ICarterModule).Assembly), RuntimeInformation.FrameworkDescription, "", Boot.Ms));
    }

    /// <summary>The version the restore resolved, without the source revision after the plus.</summary>
    private static string Version(Assembly assembly)
    {
        string? informational = assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
        int plus = informational?.IndexOf('+') ?? -1;
        return plus > 0 ? informational![..plus] : informational ?? assembly.GetName().Version?.ToString() ?? "";
    }
}
